src/librustc/ty/subst.rs

```diff
@@ -379,6 +380,72 @@ impl<'a, 'tcx> InternalSubsts<'tcx> {
     pub fn truncate_to(&self, tcx: TyCtxt<'tcx>, generics: &ty::Generics) -> SubstsRef<'tcx> {
         tcx.mk_substs(self.iter().take(generics.count()).cloned())
     }
+
+    /// Divides the closure substs into their respective
```

Comment:
Hmm so I'm of two minds about this change. On the *one hand*, I do think we want to move to the same 'overall structure' of separating out a set of type arguments, and this makes sense from that perspective.

On the *other hand*, I don't love just dumping all these methods that are specific to closure substs onto any old `SubstsRef`.

I'm debating what we might do about it.

Reply:
One option might be to have a method like `closure()` that can be used to access these methods, so that at least they are "namespaced". That means you would do something like this

```rust
ty::Closure(def_id, substs) => substs.closure().split()
```

The disadvantage of this is that you could invoke such a method on *any* set of substitutions. 

This method would be like:

```rust
impl SubstsRef<'tcx> {
    fn closure(self) -> ClosureSubsts<'tcx> { ClosureSubsts { params: self } }
}

impl ClosureSubsts<'tcx> { ... /* as before */ ... }
```

Reply:
I can't think of a much more type-safe option. I bet this would work out pretty well in practice. The only other alternative I can think of would be to make it so that the *first half* of the closure pair (the `def-id`) is some kind of new-type that can be used to access the substs, and it just doesn't seem like much of a win to me.